File: packages/contracts-bedrock/scripts/differential-testing.ts
import { BigNumber, utils, constants } from 'ethers'
import {
  decodeVersionedNonce,
  hashCrossDomainMessage,
  DepositTx,
  SourceHashDomain,
  encodeCrossDomainMessage,
  hashWithdrawal,
  hashOutputRootProof,
} from '@eth-optimism/core-utils'
import { SecureTrie } from '@ethereumjs/trie'
import { Account, Address, toBuffer, bufferToHex } from '@ethereumjs/util'

import { predeploys } from '../src'

const { hexZeroPad, keccak256 } = utils

const args = process.argv.slice(2)
const command = args[0]

;(async () => {
  switch (command) {
    case 'decodeVersionedNonce': {
      const input = BigNumber.from(args[1])
      const { nonce, version } = decodeVersionedNonce(input)

      const output = utils.defaultAbiCoder.encode(
        ['uint256', 'uint256'],
        [nonce.toHexString(), version.toHexString()]
      )
      process.stdout.write(output)
      break
    }
    case 'encodeCrossDomainMessage': {
      const nonce = BigNumber.from(args[1])
      const sender = args[2]
      const target = args[3]
      const value = BigNumber.from(args[4])
      const gasLimit = BigNumber.from(args[5])
      const data = args[6]

      const encoding = encodeCrossDomainMessage(
        nonce,
        sender,
        target,
        value,
        gasLimit,
        data
      )

      const output = utils.defaultAbiCoder.encode(['bytes'], [encoding])
      process.stdout.write(output)
      break
    }
    case 'hashCrossDomainMessage': {
      const nonce = BigNumber.from(args[1])
      const sender = args[2]
      const target = args[3]
      const value = BigNumber.from(args[4])
      const gasLimit = BigNumber.from(args[5])
      const data = args[6]

      const hash = hashCrossDomainMessage(
        nonce,
        sender,
        target,
        value,
        gasLimit,
        data
      )
      const output = utils.defaultAbiCoder.encode(['bytes32'], [hash])
      process.stdout.write(output)
      break
    }
    case 'hashDepositTransaction': {
      // The solidity transaction hash computation currently only works with
      // user deposits. System deposit transaction hashing is not supported.
      const l1BlockHash = args[1]
      const logIndex = BigNumber.from(args[2])
      const from = args[3]
      const to = args[4]
      const mint = BigNumber.from(args[5])
      const value = BigNumber.from(args[6])
      const gas = BigNumber.from(args[7])
      const data = args[8]

      const tx = new DepositTx({
        l1BlockHash,
        logIndex,
        from,
        to,
        mint,
        value,
        gas,
        data,
        isSystemTransaction: false,
        domain: SourceHashDomain.UserDeposit,
      })

      const digest = tx.hash()
      const output = utils.defaultAbiCoder.encode(['bytes32'], [digest])
      process.stdout.write(output)
      break
    }
    case 'encodeDepositTransaction': {
      const from = args[1]
      const to = args[2]
      const value = BigNumber.from(args[3])
      const mint = BigNumber.from(args[4])
      const gasLimit = BigNumber.from(args[5])
      const isCreate = args[6] === 'true' ? true : false
      const data = args[7]
      const l1BlockHash = args[8]
      const logIndex = BigNumber.from(args[9])

      const tx = new DepositTx({
        from,
        to: isCreate ? null : to,
        value,
        mint,
        gas: gasLimit,
        data,
        l1BlockHash,
        logIndex,
        domain: SourceHashDomain.UserDeposit,
      })

      const raw = tx.encode()
      const output = utils.defaultAbiCoder.encode(['bytes'], [raw])
      process.stdout.write(output)
      break
    }
    case 'hashWithdrawal': {
      const nonce = BigNumber.from(args[1])
      const sender = args[2]
      const target = args[3]
      const value = BigNumber.from(args[4])
      const gas = BigNumber.from(args[5])
      const data = args[6]

      const hash = hashWithdrawal(nonce, sender, target, value, gas, data)
      const output = utils.defaultAbiCoder.encode(['bytes32'], [hash])
      process.stdout.write(output)
      break
    }
    case 'hashOutputRootProof': {
      const version = hexZeroPad(BigNumber.from(args[1]).toHexString(), 32)
      const stateRoot = hexZeroPad(BigNumber.from(args[2]).toHexString(), 32)
      const messagePasserStorageRoot = hexZeroPad(
        BigNumber.from(args[3]).toHexString(),
        32
      )
      const latestBlockhash = hexZeroPad(
        BigNumber.from(args[4]).toHexString(),
        32
      )

      const hash = hashOutputRootProof({
        version,
        stateRoot,
        messagePasserStorageRoot,
        latestBlockhash,
      })
      const output = utils.defaultAbiCoder.encode(['bytes32'], [hash])
      process.stdout.write(output)
      break
    }
    case 'getProveWithdrawalTransactionInputs': {
      const nonce = BigNumber.from(args[1])
      const sender = args[2]
      const target = args[3]
      const value = BigNumber.from(args[4])
      const gas = BigNumber.from(args[5])
      const data = args[6]

      // Compute the withdrawalHash
      const withdrawalHash = hashWithdrawal(
        nonce,
        sender,
        target,
        value,
        gas,
        data
      )

      // Compute the storage slot the withdrawalHash will be stored in
      const slot = utils.defaultAbiCoder.encode(
        ['bytes32', 'bytes32'],
        [withdrawalHash, utils.hexZeroPad('0x', 32)]
      )
      const key = keccak256(slot)

      // Create the account storage trie
      const storage = new SecureTrie()
      // Put a bool "true" into storage
      await storage.put(toBuffer(key), toBuffer('0x01'))

      // Put the storage root into the L2ToL1MessagePasser storage
      const address = Address.fromString(predeploys.L2ToL1MessagePasser)
      const account = Account.fromAccountData({
        nonce: 0,
        balance: 0,
        stateRoot: storage.root,
      })

      const world = new SecureTrie()
      await world.put(address.toBuffer(), account.serialize())

      const proof = await SecureTrie.createProof(storage, toBuffer(key))

      const outputRoot = hashOutputRootProof({
        version: constants.HashZero,
        stateRoot: bufferToHex(world.root),
        messagePasserStorageRoot: bufferToHex(storage.root),
        latestBlockhash: constants.HashZero,
      })

      const output = utils.defaultAbiCoder.encode(
        ['bytes32', 'bytes32', 'bytes32', 'bytes32', 'bytes[]'],
        [world.root, storage.root, outputRoot, withdrawalHash, proof]
      )
      process.stdout.write(output)
      break
    }
  }
})().catch((err: Error) => {
  console.error(err)
  process.stdout.write('')
})

File: packages/contracts-bedrock/scripts/differential-testing/differential-testing.go
package main

import (
	"bytes"
	"fmt"
	"math/big"
	"os"

	"github.com/ethereum-optimism/optimism/op-chain-ops/crossdomain"
	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/common/hexutil"
	"github.com/ethereum/go-ethereum/core/rawdb"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/trie"
)

// ABI types
var (
	// Plain dynamic bytesAbi type
	bytesAbi, _ = abi.NewType("bytes", "bytes", []abi.ArgumentMarshaling{
		{Name: "data", Type: "bytes"},
	})
	bytesArgs = abi.Arguments{
		{Type: bytesAbi},
	}

	// Plain fixed bytes32 type
	fixedBytes, _ = abi.NewType("bytes32", "bytes32", []abi.ArgumentMarshaling{
		{Name: "data", Type: "bytes32"},
	})
	fixedBytesArgs = abi.Arguments{
		{Type: fixedBytes},
	}

	// Decoded nonce tuple (nonce, version)
	decodedNonce, _ = abi.NewType("tuple", "DecodedNonce", []abi.ArgumentMarshaling{
		{Name: "nonce", Type: "uint256"},
		{Name: "version", Type: "uint256"},
	})
	decodedNonceArgs = abi.Arguments{
		{Name: "nonce", Type: decodedNonce},
	}

	// WithdrawalHash slot tuple (bytes32, bytes32)
	withdrawalSlot, _ = abi.NewType("tuple", "ArraySlotHash", []abi.ArgumentMarshaling{
		{Name: "withdrawalHash", Type: "bytes32"},
		{Name: "zeroPadding", Type: "bytes32"},
	})
	withdrawalSlotArgs = abi.Arguments{
		{Name: "slotHash", Type: withdrawalSlot},
	}

	// Withdrawal transaction tuple (uint256, address, address, uint256, uint256, bytes)
	withdrawalTransaction, _ = abi.NewType("tuple", "WithdrawalTransaction", []abi.ArgumentMarshaling{
		{Name: "nonce", Type: "uint256"},
		{Name: "sender", Type: "address"},
		{Name: "target", Type: "address"},
		{Name: "value", Type: "uint256"},
		{Name: "gasLimit", Type: "uint256"},
		{Name: "data", Type: "bytes"},
	})
	withdrawalTransactionArgs = abi.Arguments{
		{Name: "withdrawalTx", Type: withdrawalTransaction},
	}

	// Output root proof tuple (bytes32, bytes32, bytes32, bytes32)
	outputRootProof, _ = abi.NewType("tuple", "OutputRootProof", []abi.ArgumentMarshaling{
		{Name: "version", Type: "bytes32"},
		{Name: "stateRoot", Type: "bytes32"},
		{Name: "messagePasserStorageRoot", Type: "bytes32"},
		{Name: "latestBlockHash", Type: "bytes32"},
	})
	outputRootProofArgs = abi.Arguments{
		{Name: "proof", Type: outputRootProof},
	}

	// Prove withdrawal inputs tuple (bytes32, bytes32, bytes32, bytes32, bytes[])
	proveWithdrawalInputs, _ = abi.NewType("tuple", "ProveWithdrawalInputs", []abi.ArgumentMarshaling{
		{Name: "worldRoot", Type: "bytes32"},
		{Name: "storageRoot", Type: "bytes32"},
		{Name: "outputRoot", Type: "bytes32"},
		{Name: "withdrawalHash", Type: "bytes32"},
		{Name: "proof", Type: "bytes[]"},
	})
	proveWithdrawalInputsArgs = abi.Arguments{
		{Name: "inputs", Type: proveWithdrawalInputs},
	}
)

func main() {
	args := os.Args[1:]

	// This command requires arguments
	if len(args) == 0 {
		panic("Error: No arguments provided")
	}

	switch args[0] {
	case "decodeVersionedNonce":
		// Parse input arguments
		input, ok := new(big.Int).SetString(args[1], 10)
		checkOk(ok)

		// Decode versioned nonce
		nonce, version := crossdomain.DecodeVersionedNonce(input)

		// ABI encode output
		packArgs := struct {
			Nonce   *big.Int
			Version *big.Int
		}{
			nonce,
			version,
		}
		packed, err := decodedNonceArgs.Pack(&packArgs)
		checkErr(err, fmt.Sprintf("Error encoding output: %s", err))

		fmt.Print(hexutil.Encode(packed))
		break
	case "encodeCrossDomainMessage":
		// Parse input arguments
		nonce, ok := new(big.Int).SetString(args[1], 10)
		checkOk(ok)
		sender := common.HexToAddress(args[2])
		target := common.HexToAddress(args[3])
		value, ok := new(big.Int).SetString(args[4], 10)
		checkOk(ok)
		gasLimit, ok := new(big.Int).SetString(args[5], 10)
		checkOk(ok)
		data := common.FromHex(args[6])

		// Encode cross domain message
		encoded, err := encodeCrossDomainMessage(nonce, sender, target, value, gasLimit, data)
		checkErr(err, fmt.Sprintf("Error encoding cross domain message: %s", err))

		// Pack encoded cross domain message
		packed, err := bytesArgs.Pack(&encoded)
		checkErr(err, fmt.Sprintf("Error encoding output: %s", err))

		fmt.Print(hexutil.Encode(packed))
		break
	case "hashCrossDomainMessage":
		// Parse input arguments
		nonce, ok := new(big.Int).SetString(args[1], 10)
		checkOk(ok)
		sender := common.HexToAddress(args[2])
		target := common.HexToAddress(args[3])
		value, ok := new(big.Int).SetString(args[4], 10)
		checkOk(ok)
		gasLimit, ok := new(big.Int).SetString(args[5], 10)
		checkOk(ok)
		data := common.FromHex(args[6])

		// Encode cross domain message
		encoded, err := encodeCrossDomainMessage(nonce, sender, target, value, gasLimit, data)
		checkErr(err, fmt.Sprintf("Error encoding cross domain message: %s", err))

		// Hash encoded cross domain message
		hash := crypto.Keccak256Hash(encoded)

		// Pack hash
		packed, err := fixedBytesArgs.Pack(&hash)
		checkErr(err, fmt.Sprintf("Error encoding output: %s", err))

		fmt.Print(hexutil.Encode(packed))
		break
	case "hashDepositTransaction":
		// Parse input arguments
		l1BlockHash := common.HexToHash(args[1])
		logIndex, ok := new(big.Int).SetString(args[2], 10)
		checkOk(ok)
		from := common.HexToAddress(args[3])
		to := common.HexToAddress(args[4])
		mint, ok := new(big.Int).SetString(args[5], 10)
		checkOk(ok)
		value, ok := new(big.Int).SetString(args[6], 10)
		checkOk(ok)
		gasLimit, ok := new(big.Int).SetString(args[7], 10)
		checkOk(ok)
		data := common.FromHex(args[8])

		// Create deposit transaction
		depositTx := makeDepositTx(from, to, value, mint, gasLimit, false, data, l1BlockHash, logIndex)

		// RLP encode deposit transaction
		encoded, err := types.NewTx(&depositTx).MarshalBinary()
		checkErr(err, fmt.Sprintf("Error encoding deposit transaction: %s", err))

		// Hash encoded deposit transaction
		hash := crypto.Keccak256Hash(encoded)

		// Pack hash
		packed, err := fixedBytesArgs.Pack(&hash)
		checkErr(err, fmt.Sprintf("Error encoding output: %s", err))

		fmt.Print(hexutil.Encode(packed))
		break
	case "encodeDepositTransaction":
		// Parse input arguments
		from := common.HexToAddress(args[1])
		to := common.HexToAddress(args[2])
		value, ok := new(big.Int).SetString(args[3], 10)
		checkOk(ok)
		mint, ok := new(big.Int).SetString(args[4], 10)
		checkOk(ok)
		gasLimit, ok := new(big.Int).SetString(args[5], 10)
		checkOk(ok)
		isCreate := args[6] == "true"
		data := common.FromHex(args[7])
		l1BlockHash := common.HexToHash(args[8])
		logIndex, ok := new(big.Int).SetString(args[9], 10)
		checkOk(ok)

		depositTx := makeDepositTx(from, to, value, mint, gasLimit, isCreate, data, l1BlockHash, logIndex)

		// RLP encode deposit transaction
		encoded, err := types.NewTx(&depositTx).MarshalBinary()
		checkErr(err, fmt.Sprintf("Failed to RLP encode deposit transaction: %s", err))
		// Pack rlp encoded deposit transaction
		packed, err := bytesArgs.Pack(&encoded)
		checkErr(err, fmt.Sprintf("Error encoding output: %s", err))

		fmt.Print(hexutil.Encode(packed))
		break
	case "hashWithdrawal":
		// Parse input arguments
		nonce, ok := new(big.Int).SetString(args[1], 10)
		checkOk(ok)
		sender := common.HexToAddress(args[2])
		target := common.HexToAddress(args[3])
		value, ok := new(big.Int).SetString(args[4], 10)
		checkOk(ok)
		gasLimit, ok := new(big.Int).SetString(args[5], 10)
		checkOk(ok)
		data := common.FromHex(args[6])

		// Hash withdrawal
		hash, err := hashWithdrawal(nonce, sender, target, value, gasLimit, data)
		checkErr(err, fmt.Sprintf("Error hashing withdrawal: %s", err))

		// Pack hash
		packed, err := fixedBytesArgs.Pack(&hash)
		checkErr(err, fmt.Sprintf("Error encoding output: %s", err))

		fmt.Print(hexutil.Encode(packed))
		break
	case "hashOutputRootProof":
		// Parse input arguments
		version := common.HexToHash(args[1])
		stateRoot := common.HexToHash(args[2])
		messagePasserStorageRoot := common.HexToHash(args[3])
		latestBlockHash := common.HexToHash(args[4])

		// Hash the output root proof
		hash, err := hashOutputRootProof(version, stateRoot, messagePasserStorageRoot, latestBlockHash)

		// Pack hash
		packed, err := fixedBytesArgs.Pack(&hash)
		checkErr(err, fmt.Sprintf("Error encoding output: %s", err))

		fmt.Print(hexutil.Encode(packed))
		break
	case "getProveWithdrawalTransactionInputs":
		// Parse input arguments
		nonce, ok := new(big.Int).SetString(args[1], 10)
		checkOk(ok)
		sender := common.HexToAddress(args[2])
		target := common.HexToAddress(args[3])
		value, ok := new(big.Int).SetString(args[4], 10)
		checkOk(ok)
		gasLimit, ok := new(big.Int).SetString(args[5], 10)
		checkOk(ok)
		data := common.FromHex(args[6])

		wdHash, err := hashWithdrawal(nonce, sender, target, value, gasLimit, data)
		checkErr(err, fmt.Sprintf("Error hashing withdrawal: %s", err))

		// Compute the storage slot the withdrawalHash will be stored in
		slot := struct {
			WithdrawalHash common.Hash
			ZeroPadding    common.Hash
		}{
			WithdrawalHash: wdHash,
			ZeroPadding:    common.Hash{},
		}
		packed, err := withdrawalSlotArgs.Pack(&slot)
		checkErr(err, fmt.Sprintf("Error packing withdrawal slot: %s", err))

		// Compute the storage slot the withdrawalHash will be stored in
		hash := crypto.Keccak256Hash(packed)

		// Create a secure trie for storage
		storage, err := trie.NewStateTrie(
			trie.TrieID(common.HexToHash("0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")),
			trie.NewDatabase(rawdb.NewMemoryDatabase()),
		)
		checkErr(err, fmt.Sprintf("Error creating secure trie: %s", err))

		// Put a "true" bool in the storage slot
		storage.Update(hash.Bytes(), []byte{0x01})

		// Create a secure trie for the world state
		world, err := trie.NewStateTrie(
			trie.TrieID(common.HexToHash("0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")),
			trie.NewDatabase(rawdb.NewMemoryDatabase()),
		)
		checkErr(err, fmt.Sprintf("Error creating secure trie: %s", err))

		// Put the storage root into the L2ToL1MessagePasser storage
		address := common.HexToAddress("0x4200000000000000000000000000000000000016")
		account := types.StateAccount{
			Nonce:   0,
			Balance: big.NewInt(0),
			Root:    storage.Hash(),
		}

		writer := new(bytes.Buffer)
		checkErr(account.EncodeRLP(writer), fmt.Sprintf("Error encoding account: %s", err))
		world.Update(address.Bytes(), writer.Bytes())

		// Get the proof
		var proof proofList
		checkErr(storage.Prove(address.Bytes(), 0, &proof), fmt.Sprintf("Error getting proof: %s", err))

		// Get the output root
		outputRoot, err := hashOutputRootProof(common.Hash{}, world.Hash(), storage.Hash(), common.Hash{})

		// Pack the output
		output := struct {
			WorldRoot      common.Hash
			StorageRoot    common.Hash
			OutputRoot     common.Hash
			WithdrawalHash common.Hash
			Proof          [][]byte
		}{
			WorldRoot:      world.Hash(),
			StorageRoot:    storage.Hash(),
			OutputRoot:     outputRoot,
			WithdrawalHash: wdHash,
			Proof:          proof,
		}
		packed, err = proveWithdrawalInputsArgs.Pack(&output)
		checkErr(err, fmt.Sprintf("Error encoding output: %s", err))

		// Print the output
		fmt.Print(hexutil.Encode(packed[32:]))

		break
	default:
		panic(fmt.Errorf("Unknown command: %s", args[0]))
	}
}

// Custom type to write the generated proof to
type proofList [][]byte

func (n *proofList) Put(key []byte, value []byte) error {
	*n = append(*n, value)
	return nil
}

func (n *proofList) Delete(key []byte) error {
	panic("not supported")
}
